bench/bin/throughput.rs

```diff
@@ -92,6 +99,190 @@ fn start_ingester(
+        MetricsRecorder {
+            server: tokio::spawn({
+                async move {
+                    Abortable::new(stream_agent_metrics(port, scrape_delay), abort_registration)
+                        .collect::<Vec<Sample>>()
+                        .await
+                }
+            }),
+            abort_handle,
+        }
+    }
+
+    #[allow(dead_code)]
+    pub async fn stop(self) -> Vec<Sample> {
+        self.abort_handle.abort();
+        match self.server.await {
+            Ok(data) => data,
+            Err(e) => panic!("error waiting for thread: {}", e),
+        }
+    }
+}
+
+fn data_pair_for(name: &str) -> impl Fn(&Sample) -> Option<(i64, f64)> + '_ {
+    move |s: &Sample| match (&s.value, &s.labels.get("outcome")) {
+        (Value::Untyped(raw), Some("success")) if s.metric.as_str() == name => {
+            Some((s.timestamp.timestamp_millis(), *raw))
+        }
+        _ => None,
+    }
+}
+// **** IMPORTED **************************
+
+// Sample number of lines
+fn calculate_fs_line_metrics(samples: &[Sample]) -> (f64, f64) {
+    let fs_sample_data = samples
+        .iter()
+        .filter_map(|s| match s.value {
+            Value::Counter(raw) if s.metric.as_str() == "logdna_agent_fs_lines" => Some(raw),
+            _ => None,
+        })
+        .zip(samples.iter().filter_map(|t| match t.value {
+            Value::Counter(_) if t.metric.as_str() == "logdna_agent_fs_lines" => {
+                Some(t.timestamp.timestamp_millis())
+            }
+            _ => None,
+        }))
+        .collect::<Vec<(f64, i64)>>();
+
+    let fs_total_time = (fs_sample_data.last().unwrap().1 - fs_sample_data[0].1) / 1000;
+    let fs_total_lines = fs_sample_data.last().unwrap().0;
+    let fs_lines_rate = fs_total_lines / fs_total_time as f64;
+
+    (fs_total_lines, fs_lines_rate)
+}
+
+// Sample maximum memory
+fn calculate_memory_max(samples: &[Sample]) -> f64 {
+    let mut max_value: f64 = 0.0;
+    let private_virtual_memory = samples
+        .iter()
+        .filter_map(|m| match m.value {
+            Value::Gauge(raw) if m.metric.as_str() == "process_virtual_memory_bytes" => Some(raw),
+            _ => None,
+        })
+        .collect::<Vec<f64>>();
+
+        for &val in private_virtual_memory.iter() {
+            if val > max_value {
+                max_value = val
+            }
+        }
+
+        max_value
+}
+
+// Sample ingest requests
+fn calculate_ingest_time_metrics(samples: &[Sample]) -> (i64, f64) {
+    let ingest_duration_sample = samples
+        .iter()
+        .filter_map(data_pair_for(
+            "logdna_agent_ingest_request_duration_seconds_sum",
+        ))
+        .zip(samples.iter().filter_map(data_pair_for(
+            "logdna_agent_ingest_request_duration_seconds_count",
+        )))
+        .collect::<Vec<((i64, f64), (i64, f64))>>();
+
+    let ingest_total_time = (ingest_duration_sample.last().unwrap().0.0 - ingest_duration_sample[0].0.0)/1000;
```

Comment:
FYI: You could use the rust destructuring operation to pull apart the tuples and give them more meaningful names than `.0`. 
```
let ((last_sum_ts, last_sum_val), (last_count_ts, last_count_val)) = ingest_duration_sample.last().unwrap();
```
The destructuring construct is pretty powerful and works on deeply nested data. It's one of the cooler features of rust that is missing in langs like Java or C++.

Reply:
This was very helpful and makes the code a lot more readable. Great suggestion!